crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,748 @@
+            "testProp": self.test_prop,
+        }))
+        .unwrap();
+
+        let signature = BASE64_URL_SAFE_NO_PAD
+            .encode(sign.sign(agent_info.as_bytes()).to_bytes());
+
+        let info = serde_json::to_string(&serde_json::json!({
+            "agentInfo": agent_info,
+            "signature": match self.signature {
+                Some(signature) => signature,
+                None => &signature,
+            }
+        }))
+        .unwrap();
+
+        let addr = format!(
+            "http://{:?}/bootstrap/{}/{}",
+            self.addr,
+            self.space_url,
+            match self.final_agent_pk {
+                Some(fapk) => fapk,
+                None => match self.agent_url {
+                    Some(agent_url) => agent_url,
+                    None => &pk,
+                },
+            },
+        );
+
+        match ureq::put(&addr).send_string(&info) {
+            Ok(res) => {
+                let res = res.into_string()?;
+                if res != "{}" {
+                    return Err(std::io::Error::other("InvalidResponse"));
+                }
+                Ok(PutInfoRes { info, agent: pk })
+            }
+            Err(ureq::Error::Status(status, res)) => {
+                let res = res.into_string()?;
+                Err(std::io::Error::other(format!("status {status}: {res}")))
+            }
+            Err(ureq::Error::Transport(err)) => Err(std::io::Error::other(err)),
+        }
+    }
+}
+
+#[test]
+fn happy_bootstrap_put_get() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let PutInfoRes { info, .. } = PutInfo {
+        addr: s.listen_addr(),
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    println!("{addr}");
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    println!("{res}");
+
+    // make sure it is valid json and only contains one entry
+    let r: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+    assert_eq!(1, r.len());
+
+    // make sure that info byte-wise matches our put
+    assert_eq!(format!("[{info}]"), res);
+}
+
+#[test]
+fn happy_empty_server_health() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+    let addr = format!("http://{:?}/health", s.listen_addr());
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    assert_eq!("{}", res);
+}
+
+#[test]
+fn happy_empty_server_bootstrap_get() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    assert_eq!("[]", res);
+}
+
+#[test]
+fn tombstone_will_not_put() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let _ = PutInfo {
+        addr: s.listen_addr(),
+        is_tombstone: true,
+        ..Default::default()
+    }
+    .call()
```

Comment:
Return an error so that the caller knows it was unsuccessful? This would go into the server implementation obviously.
Checking there, `call` should return an error here, because `created_at` is too far in the past. Yet the result to `call` is unwrapped. What's happening?

Reply:
Did you write this comment in the right test? The created_at here is valid. A correct tombstone just ensures there are no entries in the bootstrap server for this agent. There are none in this case already, but that's still a success here.

Reply:
Ah, I had misunderstood that. All good.